### PR TITLE
[Multimodal] Update version bounds

### PR DESCRIPTION
**Pull request by sxjscience**
*Issue #, if available:*

https://github.com/autogluon/autogluon/issues/612

*Description of changes:*

Update the version bounds in multimodal dependencies.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by sxjscience** [comment]
For version bounds, we can adopt the guideline to always use:

```python
# Packages that have low--up bounds.
PACKAGE_NAME>=LOWER_BOUND,<UPPER_BOUND

# Packages that must be upgraded if it's lower version
PACKAGE_NAME>=LOWER_BOUND

# Most common packages
PACKAGE_NAME
```

File: multimodal/setup.py [hunk 1]
```diff
@@ -28,35 +28,33 @@
...
     f"autogluon.common=={version}",
-    "pytorch-metric-learning>=1.3.0,<1.4.0",
+    "pytorch-metric-learning>=1.3.0,<2.0",
     "nlpaug>=1.1.10,<1.2.0",
     "nltk>=3.4.5,<4.0.0",
-    "openmim>0.1.5,<=0.2.1",
-    "defusedxml>=0.7.1,<=0.7.1",
-    "albumentations>=1.1.0,<=1.2.0",
-    "jinja2>=3.0.3,<3.1",
+    "openmim>0.1.5,<0.4.0",
```
Comment (sxjscience) [inline multimodal/setup.py:53]: @bryanyzhu I updated the version bound associated with openmim.

File: multimodal/setup.py [hunk 2]
```diff
@@ -28,35 +28,33 @@
...
-    "omegaconf>=2.1.1,<2.2.0",
+    "transformers>=4.23.0,<4.27.0",
+    "nptyping>=1.4.4,<2.5.0",
+    "omegaconf>=2.1.1,<2.3.0",
     "sentencepiece>=0.1.95,<0.2.0",
     f"autogluon.core[raytune]=={version}",
     f"autogluon.features=={version}",
     f"autogluon.common=={version}",
-    "pytorch-metric-learning>=1.3.0,<1.4.0",
+    "pytorch-metric-learning>=1.3.0,<2.0",
```
Comment (sxjscience) [inline multimodal/setup.py:50]: @zhiqiangdon I noticed that pytorch-metric-learning released version 2.0. Do we need to consider to update the version?

**Comment by Innixma** [comment]
@sxjscience Regarding this: how about the following:

```python
"numpy",  # version range defined in `_setup_utils.py`
"pandas",  # version range defined in `_setup_utils.py`
"torchvision>=foo,<bar",

...

```

File: multimodal/setup.py [hunk 3]
```diff
@@ -28,35 +28,33 @@
...
     "nlpaug>=1.1.10,<1.2.0",
     "nltk>=3.4.5,<4.0.0",
-    "openmim>0.1.5,<=0.2.1",
-    "defusedxml>=0.7.1,<=0.7.1",
-    "albumentations>=1.1.0,<=1.2.0",
-    "jinja2>=3.0.3,<3.1",
+    "openmim>0.1.5,<0.4.0",
+    "defusedxml>=0.7.1,<0.7.2",
+    "jinja2>=3.0.3,<3.2",
```
Comment (Innixma) [inline multimodal/setup.py:55]: jinja2 needs to be `<3.1` or it will error, at least it did in my PR before I dropped to `<3.1`.

Comment (Innixma) [reply multimodal/setup.py:55]: The error may have been from the `requirements_docs` dependencies though, so maybe `<3.2` is ok for the package depencenies, lets keep it `<3.2` and see if it passes

File: multimodal/setup.py (hunk 1 above)
Comment (bryanyzhu) [reply multimodal/setup.py:53]: Cool, this should be fine.

File: multimodal/setup.py (hunk 2 above)
Comment (h-vetinari) [reply multimodal/setup.py:50]: I [checked](https://github.com/conda-forge/autogluon.multimodal-feedstock/pull/15/commits/8164e39502a87c0ea265145340d70e4d0e6f5553) and `<2` would be enough to get things working. It's not ideal and should be lifted to allow v2, but it's less urgent than lifting the cap from `<1.4`.

File: multimodal/setup.py [hunk 4]
```diff
@@ -20,43 +20,41 @@
...
+    "Pillow",  # version range defined in `core/_setup_utils.py`
+    "tqdm",  # version range defined in `core/_setup_utils.py`
     "boto3",
     "requests",
-    "jsonschema<=4.8.0",
-    "seqeval<=1.2.2",
-    "evaluate<=0.3.0",
-    "accelerate>=0.9,<0.16",
+    "jsonschema<4.18",
+    "seqeval<1.2.3",
```
Comment (h-vetinari) [inline multimodal/setup.py:32]: Unless you expect patch versions to be breaking, I'd suggest
```suggestion
    "seqeval<1.3",
```

File: multimodal/setup.py [hunk 5]
```diff
@@ -20,43 +20,41 @@
...
     "timm<0.7.0",
     "torch>=1.9,<1.14",
     "torchvision<0.15.0",
     "fairscale>=0.4.5,<0.4.14",
     "scikit-image>=0.19.1,<0.20.0",
-    "smart_open>=5.2.1,<5.3.0",
     "pytorch_lightning>=1.8.0,<1.10.0",
-    "text-unidecode<=1.3",
+    "text-unidecode<1.4",
     "torchmetrics>=0.8.0,<0.9.0",
```
Comment (h-vetinari) [inline multimodal/setup.py:42]: Latest version for this is 0.11.1, could the cap be lifted?

File: multimodal/setup.py [hunk 6]
```diff
@@ -20,43 +20,41 @@
...
-    "seqeval<=1.2.2",
-    "evaluate<=0.3.0",
-    "accelerate>=0.9,<0.16",
+    "jsonschema<4.18",
+    "seqeval<1.2.3",
+    "evaluate<0.4.0",
+    "accelerate>=0.9,<0.17",
     "timm<0.7.0",
     "torch>=1.9,<1.14",
     "torchvision<0.15.0",
```
Comment (h-vetinari) [inline multimodal/setup.py:37]: Side note: what happened to the torchtext dependency? 🤔

Comment (sxjscience) [reply multimodal/setup.py:37]: We managed to remove it by upgrading lightning... See https://github.com/autogluon/autogluon/pull/2739

File: multimodal/setup.py (hunk 5 above)
Comment (sxjscience) [reply multimodal/setup.py:42]: @zhiqiangdon Can we list the cap of torchmetrics?

File: multimodal/setup.py (hunk 4 above)
Comment (sxjscience) [reply multimodal/setup.py]: Thanks. Revised.

File: multimodal/setup.py (hunk 3 above)
Comment (sxjscience) [reply multimodal/setup.py:55]: Okay, let's see if the CI + document build can pass.

File: multimodal/src/autogluon/multimodal/data/process_text.py [hunk 7]
```diff
@@ -235,7 +235,7 @@ def collate_fn(self, text_column_names: Optional[List] = None) -> Dict:
 
     def build_one_token_sequence(
         self,
-        text_tokens: Dict[str, NDArray[(Any,), np.int32]],
+        text_tokens: Dict[str, NDArray],
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/data/process_text.py:238]: @zhiqiangdon @cheungdaven I changed the typehints to be just `NDArray` due to the API change of the latest nptyping.

File: multimodal/setup.py (hunk 1 above)
Comment (sxjscience) [reply multimodal/setup.py:53]: Thanks!

File: multimodal/setup.py (hunk 5 above)
Comment (zhiqiangdon) [reply multimodal/setup.py:42]: We can't loose the cap of torchmetrics for now. We need to make some changes in our data preprocessing to loos the cap.

File: multimodal/setup.py (hunk 2 above)
Comment (zhiqiangdon) [reply multimodal/setup.py:50]: pytorch-metric-learning seems okay to upgrade.

File: multimodal/src/autogluon/multimodal/data/process_text.py (hunk 7 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/data/process_text.py:238]: `NDArray ` should be fine, but the shape and dtype info are no longer available.

**Review by zhiqiangdon — APPROVED** [review]
LGTM!

**Pull request merged**